### PR TITLE
Add unit tests

### PR DESCRIPTION
**Pull request by ale8k**
Added unit tests to both services and both controllers.
Used supertest for endpoint testing.

### CODE REVIEW

File: backend/package.json [hunk 1]
```diff
@@ -10,29 +10,31 @@
     "start": "node dist/Server.js",
     "test:unit": "jest unit --detectOpenHandles --colors",
     "test:int": "jest int --detectOpenHandles --colors",
-    "test:dev:unit": "jest --watch unit --detectOpenHandles --colors"
+    "test:unit:watch": "jest --watch unit --detectOpenHandles --colors"
```
Comment (rakeshshubhu) [inline backend/package.json:13]: these looks weird tbh. you have repeating `test:dev:unit` and `test:unit`. 
Just to make our life easier can we just write `test` & `test:watch` for `jest unit` where your integration test script will be the same.

File: backend/src/App.ts [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+                            CMSAuthService.db = client.db(process.env.DB_NAME as string);
+                            console.log("CMSAuthService db reference stored...");
+                            DbUpdateService.db = client.db(process.env.DB_NAME as string);
+                            console.log("DbUpdateService db reference stored...");
+                            console.log("Resolving server...");
+                            resolve(this.APP);
+                        }
+                    }
+                );
+            }));
```
Comment (rakeshshubhu) [inline backend/src/App.ts:74]: I'm so confused with this one. Looks fine but is there any way we could've achieved it effectively with the use of `Promise.all()`

File: backend/src/controllers/CMSTutorialController.ts [hunk 3]
```diff
@@ -20,7 +20,7 @@ export default class CMSTutorialController {
      * @param {Response} res our res obj
      */
     public static getAllTutorials(req: Request, res: Response): void {
-        TutorialUpdateService.getAllDocuments<Tutorial>("tutorials").then(
+        DbUpdateService.getAllDocuments<Tutorial>("tutorials").then(
```
Comment (rakeshshubhu) [inline backend/src/controllers/CMSTutorialController.ts:23]: Lot better.

File: backend/src/controllers/CMSTutorialController.ts [hunk 4]
```diff
@@ -41,8 +41,8 @@ export default class CMSTutorialController {
             const { name, html, markdown, category } = req.body;
             const dto = new Tutorial(name, html, markdown, category, userId, username, true);
 
-            TutorialUpdateService.createDocument<Tutorial>("tutorials", dto).then(
-                () => res.status(201).send({ ok: 1, n: 1 }).end(),
+            DbUpdateService.createDocument<Tutorial>("tutorials", dto).then(
+                (result) => res.status(201).send(result).end(),
                 () => res.status(500).send(INTERNAL_ERROR_TEXT + "TODO").end()
```
Comment (rakeshshubhu) [inline backend/src/controllers/CMSTutorialController.ts:46]: Have you updated the comments with todo as well?

File: backend/src/controllers/__tests__/CMSLoginController.unit.test.ts [hunk 5]
```diff
@@ -0,0 +1,109 @@
+import Server from "../../App";
+import supertest from "supertest";
+import { MongoClient, Db, Collection } from "mongodb";
+import { UNAUTHORISED_TEXT, BAD_REQUEST_TEXT } from "../../constants";
```
Comment (rakeshshubhu) [inline backend/src/controllers/__tests__/CMSLoginController.unit.test.ts:4]: can we set these constants in jest global? that way we dont have to import in other tests as well.

File: backend/src/controllers/__tests__/CMSLoginController.unit.test.ts [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+/**
+ * This test utilises [supertest]{@link https://www.npmjs.com/package/supertest}
+ * and initialises the full server to perform the test per test.
```
Comment (rakeshshubhu) [inline backend/src/controllers/__tests__/CMSLoginController.unit.test.ts:8]: can you update the description in a more understandable way?

File: backend/src/controllers/__tests__/CMSLoginController.unit.test.ts [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+            useUnifiedTopology: true
+        });
+
+        db = await connection.db(global.__MONGO_DB_NAME__);
+        collection = db.collection(collectionName);
+        await collection.insert(
+            {
+                username: "alex",
+                password: "$2b$12$Cp/IwyOUsKiJZTANWklBB.k4mv07lIV1gSLT4FbtsOI.eoFi2qfTu"
+            }
```
Comment (rakeshshubhu) [inline backend/src/controllers/__tests__/CMSLoginController.unit.test.ts:38]: why are passing your crendetials here? it should be here from `.env`. This way we can't move this project  to local.

File: backend/src/controllers/__tests__/CMSLoginController.unit.test.ts [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+     */
+    it("/cms/auth/login should login successfully, with 200 response", async (done) => {
+        await new Server({ path: ".env.testing" }).setupServer().then((app) => {
+            supertest(app).post("/cms/auth/login").send({ username: "alex", password: "p123" })
+                .expect(200)
+                .expect(response => {
+                    expect(response.body.successfulLogin).toBe(true);
+                }).end(done);
+        });
+    });
```
Comment (rakeshshubhu) [inline backend/src/controllers/__tests__/CMSLoginController.unit.test.ts:57]: I kind of loved the comment never seen that. 
anyway  `await` are mostly used for asynchronous behaviour and don' think its fine to use then like that even it's fine.
can you update it to something like 
```js
const json = await fetch()
const x = await json;
x.toMatch({})
```

File: backend/src/controllers/__tests__/CMSLoginController.unit.test.ts [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+
+    it("/cms/auth/login should login unsuccessfully, with 401 response", async (done) => {
+        await new Server({ path: ".env.testing" }).setupServer().then((app) => {
+            supertest(app).post("/cms/auth/login").send({ username: "alex", password: "p1233" })
+                .expect(401)
+                .expect(response => {
+                    expect(response.text).toBe(UNAUTHORISED_TEXT);
+                }).end(done);
+        });
+    });
```
Comment (rakeshshubhu) [inline backend/src/controllers/__tests__/CMSLoginController.unit.test.ts:67]: username and password seems repeating it self can you make it a `credentials` object before all of the unit tests?

File: backend/src/controllers/__tests__/CMSTutorialController.unit.test.ts [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+    let tutorialsCollection: Collection;
+
+    beforeAll(async () => {
+        process.env.PORT = "3003";
+        process.env.DB_URL = global.__MONGO_URI__;
+        process.env.DB_NAME = global.__MONGO_DB_NAME__;
+        process.env.BCRYPT_SALT = "12";
+        process.env.JWT_KEY = "iamakeylol";
+        process.env.JWT_EXPIRY = "2 days";
+
```
Comment (rakeshshubhu) [inline backend/src/controllers/__tests__/CMSTutorialController.unit.test.ts:29]: if these are the configs that you're gonna go with why not set a `test-configs.ts` file.

File: backend/src/controllers/__tests__/CMSTutorialController.unit.test.ts [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+                        .send(new PartialTutorial("tutorial", "html", "markdown", "category"))
+                        .expect(201).end(() => {
+                            agent.get("/cms/tutorials/all").set("Authorization", `Bearer ${jwt}`).expect((response) => {
+                                const tutArr = response.body as Tutorial[];
+                                const tut = tutArr[0];
+                                console.log(tut);
+                                expect(tut.isAvailable).toBeTruthy();
+                            }).end(done);
+                        });
+                });
```
Comment (rakeshshubhu) [inline backend/src/controllers/__tests__/CMSTutorialController.unit.test.ts:70]: this is too high level for me. havent seen done twice. No idea about supertest. Explain me that alright?

File: backend/src/services/__tests__/CMSAuthService.unit.test.ts [hunk 12]
```diff
@@ -1,5 +1,59 @@
...
+                password: "$2b$12$Cp/IwyOUsKiJZTANWklBB.k4mv07lIV1gSLT4FbtsOI.eoFi2qfTu"
+            }
+        );
+
+        testJwt = jwt.sign(
+            { username: "alex ", userId: "1" },
+            process.env.JWT_KEY as string,
+            { expiresIn: "2 days" }
+        );
+    });
```
Comment (rakeshshubhu) [inline backend/src/services/__tests__/CMSAuthService.unit.test.ts:34]: these credentials needs to be changed. Create a test on top level to test whether with a given credentials the user is connected or associated with the db.

File: backend/src/App.ts (hunk 2 above)
Comment (ale8k) [reply backend/src/App.ts:74]: Almost yeah, but promise.all doesn't force an order, instead it let's us perform a callback after *all* have been resolved. So we can Promise.all(). There's one issue though, when the DB connects we resolve the wrapper promise, without this we got no way to ensure the db sets in time before we return the .THIS.APP and if that resolves too late, then jest will think we're importing stuff after tests run. This was the only solution I could think of honestly. Try fiddle with it and if you come up with better one, let me know and we adjust this.
```
Promise.all([new Promise(res => {
    setTimeout(() => {
        console.log("hi")
        res();
    }, 10000)
}),
new Promise(res => {
    setTimeout(() => {
        console.log("hi from second")
        res();
    }, 5000)
})])
Promise {<pending>}
VM405:9 hi from second
VM405:3 hi
```
Notice the order doesn't matter? ^ This is the first problem I mentioned.

File: backend/package.json (hunk 1 above)
Comment (ale8k) [reply backend/package.json:13]: Do you mean like `"test:unit:watch": "npm run test:unit --watch"` ? Sure, didn't even notice this lol.

File: backend/src/controllers/CMSTutorialController.ts (hunk 3 above)
Comment (ale8k) [reply backend/src/controllers/CMSTutorialController.ts:23]: When I renamed it first time I didn't notice I made it for general db queries lol mybad!!

File: backend/src/controllers/CMSTutorialController.ts (hunk 4 above)
Comment (ale8k) [reply backend/src/controllers/CMSTutorialController.ts:46]: No no, the "TODO" is a diff ticket, the result sent here is proper evaluation of the result + testing is more consistent with a result directly from the service we're using, I initially hardcoded it lol.

File: backend/src/controllers/__tests__/CMSLoginController.unit.test.ts (hunk 5 above)
Comment (ale8k) [reply backend/src/controllers/__tests__/CMSLoginController.unit.test.ts:4]: We're testing against the constants the app uses, if we make them global it kinda defeats the point doesn't it? 😕

File: backend/src/controllers/__tests__/CMSLoginController.unit.test.ts (hunk 8 above)
Comment (ale8k) [reply backend/src/controllers/__tests__/CMSLoginController.unit.test.ts:57]: Yup, await isn't needed for these.  It's better off to not use it whatsoever when creating 'new Server'. Will change now.

File: backend/src/controllers/__tests__/CMSLoginController.unit.test.ts (hunk 9 above)
Comment (ale8k) [reply backend/src/controllers/__tests__/CMSLoginController.unit.test.ts:67]: Amended :)

File: backend/src/controllers/__tests__/CMSTutorialController.unit.test.ts (hunk 10 above)
Comment (ale8k) [reply backend/src/controllers/__tests__/CMSTutorialController.unit.test.ts:29]: They're being manually set at test run-time for this we'll need another lib to set jest-environment. I'll add it in

File: backend/src/controllers/__tests__/CMSTutorialController.unit.test.ts (hunk 11 above)
Comment (ale8k) [reply backend/src/controllers/__tests__/CMSTutorialController.unit.test.ts:70]: Aha sure yeah, it's performing chained HTTP requests following our app structure, so it logs in, creates one and then tries to get them all. That's all XD

File: backend/src/services/__tests__/CMSAuthService.unit.test.ts (hunk 12 above)
Comment (ale8k) [reply backend/src/services/__tests__/CMSAuthService.unit.test.ts:34]: Not sure what you mean here? These are credentials being placed into the test db to be used to login lol.

File: backend/src/controllers/__tests__/CMSTutorialController.unit.test.ts (hunk 10 above)
Comment (ale8k) [reply backend/src/controllers/__tests__/CMSTutorialController.unit.test.ts:29]: I'll shove them in an exporting function for now, think that's probably better lol.

File: backend/package.json (hunk 1 above)
Comment (rakeshshubhu) [reply backend/package.json:13]: This one resolved then.

File: backend/src/App.ts (hunk 2 above)
Comment (rakeshshubhu) [reply backend/src/App.ts:74]: Oh if order matters then promise.all wont help. Would be great to use rxjs here :P tbh.

File: backend/src/services/__tests__/CMSAuthService.unit.test.ts (hunk 12 above)
Comment (rakeshshubhu) [reply backend/src/services/__tests__/CMSAuthService.unit.test.ts:34]: yeah.. if I have have a user with alex.. then these tests are doomed. I just wanted to say I can test my db with any credentials.

File: backend/src/App.ts (hunk 2 above)
Comment (ale8k) [reply backend/src/App.ts:74]: That's a fair point actually lol. Maybe if we get everyone comfortable, we could pipe all the dependency environments into an observable XD

File: backend/src/services/__tests__/CMSAuthService.unit.test.ts (hunk 12 above)
Comment (ale8k) [reply backend/src/services/__tests__/CMSAuthService.unit.test.ts:34]: They reset every suite man

File: backend/src/controllers/__tests__/CMSLoginController.unit.test.ts (hunk 5 above)
Comment (rakeshshubhu) [reply backend/src/controllers/__tests__/CMSLoginController.unit.test.ts:4]: yeah but they're never gonna change. I mean we can import the same constants and put it inside of our jest globals.

File: backend/src/controllers/__tests__/CMSLoginController.unit.test.ts (hunk 6 above)
Comment (rakeshshubhu) [reply backend/src/controllers/__tests__/CMSLoginController.unit.test.ts:9]: @ale8k this was hidden in case you missed it.

**Comment by ale8k** [comment]
Rakesh has approved but on discord lol.

**Comment by rakeshshubhu** [comment]
aha sorry

**Pull request merged**